src/main/scala/cromwell/binding/WdlNamespace.scala

```diff
@@ -2,19 +2,109 @@ package cromwell.binding
 
 import java.io.File
 
-import cromwell.binding.WdlNamespace.ImportResolver
-import cromwell.binding.command._
 import cromwell.binding.types._
 import cromwell.binding.values.WdlValue
 import cromwell.parser.{AstTools, WdlParser}
 import cromwell.parser.WdlParser._
 import cromwell.util.FileUtil
-import AstTools.EnhancedAstNode
+import cromwell.parser.AstTools.{AstNodeName, EnhancedAstNode, EnhancedAstSeq}
 import scala.collection.JavaConverters._
 import scala.util.{Failure, Success, Try}
 import scala.language.postfixOps
 
 /**
+ * Define WdlNamespace as a sum type w/ two states - one containing a local workflow and one without.
+ * The latter is a valid state for a WDL file, however only the former can be requested to be run, so
+ * any constructs (e.g. WorkflowManagerActor) expecting to run a workflow should only take the `NamespaceWithWorkflow`
+ */
+sealed trait WdlNamespace extends WdlValue {
+  final val wdlType = WdlNamespaceType
+
+  def ast: Ast // FIXME: I think this is only used by the syntax highlighting, can it go away once we're built?
+  def importedAs: Option[String] // Used when imported with `as` 
+  def imports: Seq[Import] // FIXME: Change to Set?
+  def namespaces: Seq[WdlNamespace] // FIXME: Change to Set? FIXME: Rename to importedNamespaces?
+  def tasks: Seq[Task] // FIXME: Change to Set?
+  def importedWorkflows: Seq[Workflow] // FIXME: Change to Set?
+  def terminalMap: Map[Terminal, WdlSource]
+  def allWorkflows: Set[Workflow]
+
+  // Convenience method for findTask in the context of this namespace
+  def findTask(name: String): Option[Task] = WdlNamespace.findTask(name, namespaces, tasks)
+
+  override def toRawString = ???
+}
+
+/**
+ * A valid Namespace which doesn't have a locally defined Workflow. This should pass any validity checking but is not
+ * directly runnable by `WorkflowManagerActor`
+ */
```

Comment:
Algebraic data types are better because they enforce the invariant that there's a single runnable workflow throughout the system, instead of having `workflows` and calling `head` everywhere. This already resolved a subtle landmine I found lurking where `workflows` was defined as being imported workflows plus local workflows. Note the order considering we call `head` throughout our code.

Declaring this via a sum type lis basically making it an enum, saying "It can be exactly this or exactly that" at compile time and not have to worry that something weird happens at runtime. For instance, this way no one needs to ask "Will that head blow up?" every single code review.

Besides, don't forget to add in the loss of all of that `.head`